### ISSUE TITLE
Implement Canary-Qwen 2.5B Engine (ASR + LLM Dual Pass) 

### ISSUE THREAD

**Issue by WanderingAstronomer** [WanderingAstronomer/Vociferous#16]
## Problem Statement

Current transcription engines (WhisperTurbo, Voxtral) have accuracy and feature limitations:

**Whisper Large V3 Turbo:**
- WER: ~7.4% (real-world average)
- Speed: 216x real-time
- No built-in text refinement

**Canary-Qwen 2.5B offers:**
- WER: **5.63%** (24% improvement!)
- Speed: **418x real-time** (93% faster!)
- **Dual-mode:** ASR encoder + LLM decoder in ONE model

---

## Goals

Implement Canary-Qwen 2.5B as a new engine option with:
1. ✅ ASR mode:  Audio → raw text transcription
2. ✅ LLM mode: Text refinement (grammar, punctuation, fluency)
3. ✅ Dual-pass workflow: ASR → LLM polish in one command
4. ✅ Benchmarks showing accuracy/speed improvements
5. ✅ Documentation and examples

---

### Model Architecture

**Canary-Qwen 2.5B = FastConformer (encoder) + Qwen3-1.7B (decoder)**

- **ASR Mode:** FastConformer processes audio → text
- **LLM Mode:** Qwen3 decoder refines text → polished output

### API Research Needed

**Questions to answer:**
1. How to load the model?  (HuggingFace Transformers?  Custom loader?)
2. How to switch between ASR and LLM modes? 
3. What are the input/output formats?
4. VRAM requirements? 
5. Quantization options (int8, fp16)?

**Resources:**
- Model page: https://huggingface.co/nvidia/canary-qwen-2.5b
- Benchmarks: https://huggingface.co/spaces/hf-audio/open_asr_leaderboard
- Example code: [search for official examples]

---

### Phase 1: Model Loading

**File:** `vociferous/engines/canary_qwen.py`

```python
"""Canary-Qwen 2.5B dual-mode engine (ASR + LLM)."""

from pathlib import Path
from vociferous.domain.model import EngineConfig, TranscriptionOptions, TranscriptSegment

class CanaryQwenEngine:
    """
    Canary-Qwen 2.5B hybrid ASR-LLM engine.
    
    Modes:
      - ASR: Audio → raw transcript (FastConformer encoder)
      - LLM: Text → polished transcript (Qwen3 decoder)
    """
    
    def __init__(self, config: EngineConfig):
        self.config = config
        self.model = None  # Lazy-loaded
        self.mode = config.params.get("mode", "asr")  # "asr" or "llm"
        
        # Model parameters
        self.device = config.device or "cuda"
        self.compute_type = config.compute_type or "float16"
        
    def _load_model(self):
        """Load Canary-Qwen model."""
        if self.model is not None:
            return
        
        # TODO: Research correct loading method
        # Options: 
        # 1. HuggingFace Transformers
        # 2. NeMo toolkit (NVIDIA)
        # 3. Custom loader
        
        from transformers import AutoModel, AutoProcessor  # Example
        
        model_id = "nvidia/canary-qwen-2.5b"
        self.processor = AutoProcessor.from_pretrained(model_id)
        self.model = AutoModel.from_pretrained(
            model_id,
            device_map=self.device,
            torch_dtype=self._get_torch_dtype(),
        )
        
    def _get_torch_dtype(self):
        """Convert compute_type to torch dtype."""
        import torch
        mapping = {
            "float16":  torch.float16,
            "float32": torch.float32,
            "bfloat16": torch.bfloat16,
        }
        return mapping.get(self.compute_type, torch.float16)
```

---

### Phase 2: ASR Mode (Audio → Text)

```python
    def transcribe_file(
        self, 
        audio_path: Path, 
        options: TranscriptionOptions
    ) -> list[TranscriptSegment]:
        """Transcribe audio file (ASR mode)."""
        if self.mode != "asr":
            raise ValueError("transcribe_file requires mode='asr'")
        
        self._load_model()
        
        # Load audio
        audio_array = self._load_audio(audio_path)
        
        # Preprocess for model
        inputs = self.processor(
            audio_array,
            sampling_rate=16000,
            return_tensors="pt",
        ).to(self.device)
        
        # Generate transcription
        with torch.no_grad():
            output_ids = self.model.generate(
                **inputs,
                max_length=512,
                num_beams=1,  # Greedy decoding for speed
            )
        
        # Decode to text
        transcription = self.processor.batch_decode(
            output_ids,
            skip_special_tokens=True,
        )[0]
        
        # Convert to segments (simple version:  one segment)
        # TODO: Research if model provides timestamps
        segments = [
            TranscriptSegment(
                text=transcription.strip(),
                start_s=0.0,
                end_s=self._get_audio_duration(audio_path),
                language=options.language or "en",
                confidence=1.0,
            )
        ]
        
        return segments
```

---

### Phase 3: LLM Mode (Text Refinement)

```python
    def refine_text(
        self, 
        raw_text: str, 
        instructions: str = None
    ) -> str:
        """Polish transcript using LLM decoder (Qwen3).
        
        Args:
            raw_text: Raw transcription from ASR
            instructions: Optional refinement instructions
            
        Returns: 
            Polished transcript with proper grammar, punctuation
        """
        self._load_model()
        
        # Default refinement instructions
        if instructions is None:
            instructions = (
                "Fix any transcription errors, add proper punctuation, "
                "improve readability, and correct grammar.  "
                "Preserve the original meaning and content."
            )
        
        # Build prompt for Qwen3 decoder
        prompt = self._build_refinement_prompt(raw_text, instructions)
        
        # Generate refined text
        inputs = self.processor(
            prompt,
            return_tensors="pt",
        ).to(self.device)
        
        with torch.no_grad():
            output_ids = self.model.generate(
                **inputs,
                max_length=len(raw_text) * 2,  # Allow expansion
                temperature=0.3,  # Low temp for consistency
                do_sample=True,
            )
        
        refined = self.processor.batch_decode(
            output_ids,
            skip_special_tokens=True,
        )[0]
        
        return refined.strip()
    
    def _build_refinement_prompt(self, text: str, instructions: str) -> str:
        """Construct prompt for LLM refinement."""
        return f"""<|im_start|>system
You are a transcript editor. Your task is to polish the following raw transcript. 

Instructions:  {instructions}
<|im_end|>
<|im_start|>user
Raw transcript: 
{text}
<|im_end|>
<|im_start|>assistant
Polished transcript:
"""
```

---

### Phase 4: Dual-Pass CLI Integration

**File:** `vociferous/cli/commands/transcribe_canary.py`

```python
"""Canary-Qwen transcription command with dual-pass support."""

import typer
from pathlib import Path

app = typer.Typer()

@app.command()
def transcribe_canary(
    audio_file: Path,
    refine: bool = typer.Option(True, help="Enable LLM refinement pass"),
    output:  Path = typer.Option(None, help="Output file path"),
):
    """Transcribe audio using Canary-Qwen 2.5B (ASR + optional LLM polish)."""
    
    from vociferous.engines.canary_qwen import CanaryQwenEngine
    from vociferous.domain.model import EngineConfig, TranscriptionOptions
    
    # ASR Pass
    typer.echo("[1/2] Transcribing audio (ASR mode)...")
    asr_config = EngineConfig(
        engine_kind="canary_qwen",
        params={"mode": "asr"},
    )
    engine = CanaryQwenEngine(asr_config)
    
    segments = engine.transcribe_file(
        audio_file,
        TranscriptionOptions(language="en"),
    )
    
    raw_text = " ".join(seg.text for seg in segments)
    typer.echo(f"Raw transcript: {raw_text[: 100]}...")
    
    # LLM Refinement Pass (optional)
    if refine:
        typer. echo("[2/2] Refining transcript (LLM mode)...")
        refined_text = engine.refine_text(raw_text)
        final_text = refined_text
    else:
        final_text = raw_text
    
    # Output
    if output:
        output. write_text(final_text, encoding="utf-8")
        typer.echo(f"✓ Saved to {output}")
    else:
        typer.echo(final_text)
```

---

### Phase 5: Integration with `transcribe-full`

**File:** `vociferous/cli/commands/transcribe_full.py`

```python
@app.command()
def transcribe_full(
    audio_file: Path,
    engine: str = typer.Option("canary_qwen", help="Engine:  whisper_turbo, voxtral, canary_qwen"),
    ...
):
    """Full pipeline:  decode → vad → condense → transcribe → polish."""
    
    # Steps 1-3: Audio preprocessing (unchanged)
    decoded = decode_audio(audio_file)
    timestamps = detect_vad(decoded)
    condensed = condense_audio(decoded, timestamps)
    
    # Step 4: Transcribe with selected engine
    if engine == "canary_qwen":
        # Use Canary-Qwen with dual-pass
        segments = transcribe_with_canary(condensed, refine=True)
    elif engine == "whisper_turbo":
        segments = transcribe_with_whisper(condensed)
    else:
        ... 
    
    # Output
    ... 
```

---

### Phase 6: Benchmarking

**File:** `tests/benchmarks/test_engine_accuracy.py`

```python
"""Benchmark engine accuracy and speed."""

import time
from pathlib import Path

BENCHMARK_FILE = Path("samples/ASR_Test_30s.flac")

def test_canary_qwen_accuracy():
    """Benchmark Canary-Qwen WER and speed."""
    engine = CanaryQwenEngine(config)
    
    # Measure speed
    start = time.time()
    segments = engine.transcribe_file(BENCHMARK_FILE, options)
    elapsed = time.time() - start
    
    # Calculate metrics
    audio_duration = 30.0  # seconds
    rtf = audio_duration / elapsed  # Real-time factor
    
    # Ground truth (manually verified)
    ground_truth = "Today is Monday December 8th 2024 at exactly 8 PM..."
    predicted = " ".join(seg.text for seg in segments)
    
    wer = calculate_wer(ground_truth, predicted)
    
    print(f"Canary-Qwen Results:")
    print(f"  WER: {wer:.2%}")
    print(f"  Speed: {rtf:.1f}x real-time")
    print(f"  Elapsed: {elapsed:.2f}s for {audio_duration}s audio")
    
    # Assert performance targets
    assert wer < 0.06, f"WER too high: {wer:.2%} (expected <6%)"
    assert rtf > 400, f"Speed too slow: {rtf:.1f}x (expected >400x)"

def test_compare_engines():
    """Compare Canary-Qwen vs WhisperTurbo."""
    results = {}
    
    for engine_name in ["canary_qwen", "whisper_turbo"]:
        engine = build_engine(engine_name)
        start = time.time()
        segments = engine.transcribe_file(BENCHMARK_FILE, options)
        elapsed = time.time() - start
        
        predicted = " ".join(seg.text for seg in segments)
        wer = calculate_wer(ground_truth, predicted)
        
        results[engine_name] = {
            "wer": wer,
            "speed": 30.0 / elapsed,
            "elapsed": elapsed,
        }
    
    # Print comparison
    print("\nEngine Comparison:")
    for name, metrics in results.items():
        print(f"  {name}:")
        print(f"    WER: {metrics['wer']:.2%}")
        print(f"    Speed: {metrics['speed']:.1f}x")
        print(f"    Time: {metrics['elapsed']:.2f}s")
    
    # Assert Canary-Qwen is better
    assert results["canary_qwen"]["wer"] < results["whisper_turbo"]["wer"]
    assert results["canary_qwen"]["speed"] > results["whisper_turbo"]["speed"]
```

---

### Phase 7: Configuration

**File:** `vociferous/config/schema.py`

```python
@dataclass
class AppConfig:
    # ...  existing fields ...
    
    # Canary-Qwen settings
    canary_qwen_enabled: bool = False
    canary_qwen_refine_by_default: bool = True
    canary_qwen_refinement_instructions: str = (
        "Fix grammar, add punctuation, improve readability"
    )
```

**File:** `~/.config/vociferous/config.toml`

```toml
[canary_qwen]
enabled = true
refine_by_default = true
refinement_instructions = "Fix grammar and add proper punctuation"
```

---

### Phase 8: Documentation

**File:** `docs/engines/canary_qwen.md`

```markdown
# Canary-Qwen 2.5B Engine

## Overview

Canary-Qwen is a hybrid ASR-LLM model that: 
- Transcribes audio with 5.63% WER (vs Whisper's 7.4%)
- Runs at 418x real-time (vs Whisper's 216x)
- Includes built-in text refinement (LLM mode)

## Usage

### Basic Transcription

```bash
vociferous transcribe audio.mp3 --engine canary_qwen
```

### With LLM Refinement

```bash
vociferous transcribe audio. mp3 --engine canary_qwen --refine
```

### Manual Two-Pass

```bash
# Pass 1: ASR
vociferous transcribe audio.mp3 --engine canary_qwen --no-refine > raw.txt

# Pass 2: Refine
vociferous refine raw.txt > polished.txt
```

## Performance

| Metric | Canary-Qwen | Whisper Turbo |
|--------|-------------|---------------|
| WER | 5.63% | 7.4% |
| Speed | 418x RT | 216x RT |
| VRAM | ~3GB | ~2.5GB |
| Modes | ASR + LLM | ASR only |

## Requirements

- CUDA-capable GPU (recommended)
- 3GB+ VRAM
- PyTorch 2.0+
- Transformers library

## Configuration

Add to `~/.config/vociferous/config.toml`:

```toml
engine = "canary_qwen"

[canary_qwen]
enabled = true
refine_by_default = true
device = "cuda"
compute_type = "float16"
```
```

---

## Acceptance Criteria

- [ ] Canary-Qwen model loads successfully
- [ ] ASR mode transcribes audio
- [ ] LLM mode refines text
- [ ] Dual-pass workflow works end-to-end
- [ ] Benchmarks show improvement over Whisper: 
  - WER < 6%
  - Speed > 400x real-time
- [ ] CLI integration complete
- [ ] Documentation written
- [ ] Tests pass
- [ ] Example outputs demonstrate quality improvement

---

## Testing Strategy

1. **Unit tests:** Model loading, ASR/LLM modes in isolation
2. **Integration tests:** CLI commands, dual-pass workflow
3. **Benchmarks:** WER and speed comparisons
4. **Manual verification:** Listen to audio, read transcripts

---

## Dependencies

- Can be implemented in parallel with Issue #2 (Engine Refactoring)
- Will benefit from Issue #2's batch processing interface

---

## References

- Model: https://huggingface.co/nvidia/canary-qwen-2.5b
- Benchmarks: OpenASR Leaderboard
- Research notes: [link to earlier conversation]
